File: apps/frontend/src/protocol.ts
// The WebSocket protocol between the page and the service, both directions.
//
// Server to browser: every message the page can receive, one variant of
// `ServerMessage` per `type`, and `decodeServerMessage`, which admits a text
// frame only as one of them. The service's half is `ServerMessage` in
// `apps/backend/src/protocol.rs`. Both halves are held to the examples in
// `tests/fixtures/server-messages.json`: each one must decode here to exactly
// itself and serialize there to exactly itself, and a type without an example
// fails on both sides.
//
// Browser to server: the builders further down, one per command the service
// handles in `handle_text_frame`.

// --- Server to browser -----------------------------------------------------

/// One line of the call's transcript.
export interface TranscriptEntry {
	role: string;
	text: string;
	route: string;
	/** Seconds since the Unix epoch. */
	ts: number;
	/** The clip or typed turn a caller line came from. */
	id?: string;
}

/// A model the leg's host can run, from its catalog.
export interface ModelEntry {
	provider: string;
	model: string;
	thinks: boolean;
}

/// The answer to `hello`: the protocol version, and which optional transports
/// this socket uses.
export interface HelloAckMessage {
	type: "hello_ack";
	version: number;
	stt_streaming: boolean;
	audio_streaming: boolean;
	mse_mp3: boolean;
}

/// The answer to `ping`, echoing its nonce and time.
export interface PongMessage {
	type: "pong";
	nonce: string;
	time: number;
}

/// The turn epoch. Clips are stamped with it; a new one retires the old leg's
/// clips, audio, and screen.
export interface EpochMessage {
	type: "epoch";
	generation: number;
}

/// A transfer began starting a leg on `route`. Clips recorded until the next
/// `epoch` are addressed to that leg.
export interface CandidateMessage {
	type: "candidate";
	route: string;
	generation: number;
}

/// The candidate leg was adopted, rolled back, or rescued away.
export interface CandidateClearedMessage {
	type: "candidate_cleared";
	generation: number;
}

/// The line's status: who is on it, their model and thinking level, and which
/// of those the caller may change.
export interface StatusMessage {
	type: "status";
	route: string;
	/** Display name for whoever is on the line. */
	label: string;
	/** The model spec the leg was started with, thinking suffix included. */
	model: string;
	/** `provider/model`, without the thinking suffix. */
	model_name: string;
	/** The level the leg reported, else the one it was asked for. */
	thinking: string;
	thinking_requested: string;
	/** Whether the leg has reported its level. */
	thinking_confirmed: boolean;
	thinking_default: string;
	levels: string[];
	models: ModelEntry[];
	models_available: boolean;
	/** Why the catalog is unavailable, when it is. */
	models_diagnostic: string | null;
	model_swaps: boolean;
	projects: string[];
}

/// A clip was taken: whole, or opened as a stream (`streaming`).
export interface AcceptedMessage {
	type: "accepted";
	id: string;
	streaming?: boolean;
}

/// A streaming clip cannot continue as a stream; send it whole instead.
export interface AbandonedMessage {
	type: "abandoned";
	id: string;
	reason: string;
}

/// The final transcript of a clip, or the echo of a typed turn.
export interface TranscriptMessage {
	type: "transcript";
	id: string;
	text: string;
}

/// A transcript was steered into the turn in progress, or queued behind
/// `waiting` turns.
export interface QueuedMessage {
	type: "queued";
	id: string;
	waiting: number;
	steered: boolean;
}

/// Something the caller should see went wrong. `id` names the clip or typed
/// turn it answers, when there is one; `stale_epoch` means that clip was
/// recorded for a leg the call has since left.
export interface ErrorMessage {
	type: "error";
	id?: string;
	code?: "stale_epoch";
	message: string;
}

/// A turn was dispatched to the leg on `route`.
export interface ThinkingMessage {
	type: "thinking";
	route: string;
	waiting: number;
}

/// A tool call on the live leg started or ended (`state` is `start` or
/// `end`); `label` names the leg.
export interface ActivityMessage {
	type: "activity";
	state: string;
	tool: string;
	detail: string;
	label: string;
}

/// A turn settled with this written reply.
export interface ReplyMessage {
	type: "reply";
	text: string;
	route: string;
}

/// A line was spoken to the caller and kept in the transcript.
export interface SpokenMessage {
	type: "spoken";
	entry: TranscriptEntry;
}

/// The transcript so far, in every connection's snapshot.
export interface HistoryMessage {
	type: "history";
	entries: TranscriptEntry[];
}

/// An utterance of synthesized speech begins; its audio follows as binary
/// frames.
export interface AudioStartMessage {
	type: "audio_start";
	generation: number;
	sequence: number;
	mime: string;
	format: string;
}

/// The utterance `sequence` has no more audio.
export interface AudioDoneMessage {
	type: "audio_done";
	generation: number;
	sequence: number;
	done: boolean;
}

/// All audio for the reply to `response_id` has been sent; `success` is false
/// when some of it could not be synthesized.
export interface FinalResponseAudioClosedMessage {
	type: "final_response_audio_closed";
	response_id: string;
	generation: number;
	success: boolean;
}

/// A display action for the stage, with the delivery sequence the page
/// confirms it by. The action is validated separately, by
/// `controller/validation.ts`, against the display protocol.
export interface DisplayMessage {
	type: "display";
	action: unknown;
	seq?: number;
}

/// The agent asked for a view of the workspace.
export interface ViewMessage {
	type: "view";
	target: string;
	reason: string;
}

/// The page's `screen_state` report was applied.
export interface ScreenStateAckMessage {
	type: "screen_state_ack";
}

export type ServerMessage =
	| HelloAckMessage
	| PongMessage
	| EpochMessage
	| CandidateMessage
	| CandidateClearedMessage
	| StatusMessage
	| AcceptedMessage
	| AbandonedMessage
	| TranscriptMessage
	| QueuedMessage
	| ErrorMessage
	| ThinkingMessage
	| ActivityMessage
	| ReplyMessage
	| SpokenMessage
	| HistoryMessage
	| AudioStartMessage
	| AudioDoneMessage
	| FinalResponseAudioClosedMessage
	| DisplayMessage
	| ViewMessage
	| ScreenStateAckMessage;

// Readers: each takes a JSON value and returns it as the type it reads, or
// INVALID. `fields` is checked by the compiler against the interface it
// reads, so a field an interface declares cannot be left unread or read as
// the wrong type.

const INVALID = Symbol("invalid");
type Read<T> = (value: unknown) => T | typeof INVALID;
type Fields<T> = { [Field in keyof T]-?: Read<T[Field]> };

const string: Read<string> = (value) =>
	typeof value === "string" ? value : INVALID;
const number: Read<number> = (value) =>
	typeof value === "number" && Number.isFinite(value) ? value : INVALID;
const boolean: Read<boolean> = (value) =>
	typeof value === "boolean" ? value : INVALID;
/** Any value that is present; its own validation happens elsewhere. */
const present: Read<unknown> = (value) =>
	value === undefined ? INVALID : value;

function literal<T extends string>(expected: T): Read<T> {
	return (value) => (value === expected ? expected : INVALID);
}

function optional<T>(read: Read<T>): Read<T | undefined> {
	return (value) => (value === undefined ? undefined : read(value));
}

function nullable<T>(read: Read<T>): Read<T | null> {
	return (value) => (value === null ? null : read(value));
}

function list<T>(read: Read<T>): Read<T[]> {
	return (value) => {
		if (!Array.isArray(value)) return INVALID;
		const items: T[] = [];
		for (const item of value) {
			const itemValue = read(item);
			if (itemValue === INVALID) return INVALID;
			items.push(itemValue);
		}
		return items;
	};
}

/// Reads exactly the declared fields, leaving out an optional one that is
/// absent. Anything else the value carries is dropped.
function object<T>(fields: Fields<T>): Read<T> {
	return (value) => {
		if (value === null || typeof value !== "object" || Array.isArray(value))
			return INVALID;
		const source = value as Record<string, unknown>;
		const result: Record<string, unknown> = {};
		for (const field of Object.keys(fields) as Array<keyof T & string>) {
			const fieldValue = fields[field](source[field]);
			if (fieldValue === INVALID) return INVALID;
			if (fieldValue !== undefined) result[field] = fieldValue;
		}
		return result as T;
	};
}

const transcriptEntry = object<TranscriptEntry>({
	role: string,
	text: string,
	route: string,
	ts: number,
	id: optional(string),
});

const modelEntry = object<ModelEntry>({
	provider: string,
	model: string,
	thinks: boolean,
});

type MessageType = ServerMessage["type"];
type MessageFields<Type extends MessageType> = Fields<
	Omit<Extract<ServerMessage, { type: Type }>, "type">
>;

const MESSAGE_FIELDS: { [Type in MessageType]: MessageFields<Type> } = {
	hello_ack: {
		version: number,
		stt_streaming: boolean,
		audio_streaming: boolean,
		mse_mp3: boolean,
	},
	pong: { nonce: string, time: number },
	epoch: { generation: number },
	candidate: { route: string, generation: number },
	candidate_cleared: { generation: number },
	status: {
		route: string,
		label: string,
		model: string,
		model_name: string,
		thinking: string,
		thinking_requested: string,
		thinking_confirmed: boolean,
		thinking_default: string,
		levels: list(string),
		models: list(modelEntry),
		models_available: boolean,
		models_diagnostic: nullable(string),
		model_swaps: boolean,
		projects: list(string),
	},
	accepted: { id: string, streaming: optional(boolean) },
	abandoned: { id: string, reason: string },
	transcript: { id: string, text: string },
	queued: { id: string, waiting: number, steered: boolean },
	error: {
		id: optional(string),
		code: optional(literal("stale_epoch")),
		message: string,
	},
	thinking: { route: string, waiting: number },
	activity: { state: string, tool: string, detail: string, label: string },
	reply: { text: string, route: string },
	spoken: { entry: transcriptEntry },
	history: { entries: list(transcriptEntry) },
	audio_start: {
		generation: number,
		sequence: number,
		mime: string,
		format: string,
	},
	audio_done: { generation: number, sequence: number, done: boolean },
	final_response_audio_closed: {
		response_id: string,
		generation: number,
		success: boolean,
	},
	display: { action: present, seq: optional(number) },
	view: { target: string, reason: string },
	screen_state_ack: {},
};

/// Every `type` the page understands.
export const SERVER_MESSAGE_TYPES = Object.keys(
	MESSAGE_FIELDS,
) as readonly MessageType[];

/// A text frame from the service as the message it is, or null when it is not
/// one: not JSON, a type the page does not know, or a known type with a field
/// missing or of the wrong kind. Fields a message does not declare are
/// dropped.
export function decodeServerMessage(text: string): ServerMessage | null {
	let value: unknown;
	try {
		value = JSON.parse(text);
	} catch {
		return null;
	}
	if (value === null || typeof value !== "object" || Array.isArray(value))
		return null;
	const type = (value as { type?: unknown }).type;
	if (typeof type !== "string" || !Object.hasOwn(MESSAGE_FIELDS, type))
		return null;
	const fields = MESSAGE_FIELDS[type as MessageType] as Fields<
		Record<string, unknown>
	>;
	const read = object(fields)(value);
	return read === INVALID ? null : ({ type, ...read } as ServerMessage);
}

// --- Browser to server -----------------------------------------------------

/// The text frame that precedes a clip's audio frame.
///
/// `generation` is the server's turn epoch as of when this clip started
/// recording, and the server drops the clip if the epoch has moved on since.
/// That is what stops speech begun before a page transfer from being acted on
/// by the leg that replaced it, so the field has to survive every change to
/// this frame.
export function helloMessage(): string {
	const mse =
		typeof MediaSource !== "undefined" &&
		MediaSource.isTypeSupported?.("audio/mpeg") === true;
	return JSON.stringify({
		type: "hello",
		version: 1,
		capabilities: {
			stt_streaming: true,
			audio_streaming: mse,
			mse_mp3: mse,
		},
	});
}

export function screenStateMessage(
	view: string,
	hasVisual: boolean,
	visualKind: string | null,
	title: string,
	stale: boolean,
	generation?: number,
	pinned?: boolean,
	objectIds?: string[],
	appliedSeq?: number,
	rejected?: { seq: number; reason: string },
): string {
	return JSON.stringify({
		type: "screen_state",
		view,
		has_visual: hasVisual,
		visual_kind: visualKind,
		title,
		stale,
		...(generation !== undefined ? { generation } : {}),
		...(pinned !== undefined ? { pinned } : {}),
		...(objectIds !== undefined ? { object_ids: objectIds } : {}),
		...(appliedSeq !== undefined ? { applied_seq: appliedSeq } : {}),
		...(rejected !== undefined ? { rejected } : {}),
	});
}

export function sttStartHeader(clip: {
	id: string;
	mime: string;
	epoch: number;
}): string {
	return JSON.stringify({
		type: "stt_start",
		clip_id: clip.id,
		generation: clip.epoch,
		mime: clip.mime,
	});
}

export function sttChunkHeader(
	clip: { id: string; epoch: number },
	sequence: number,
): string {
	return JSON.stringify({
		type: "stt_chunk",
		clip_id: clip.id,
		generation: clip.epoch,
		sequence,
	});
}

export function sttEndHeader(clip: { id: string; epoch: number }): string {
	return JSON.stringify({
		type: "stt_end",
		clip_id: clip.id,
		generation: clip.epoch,
	});
}

export function sttCancelHeader(clip: { id: string; epoch: number }): string {
	return JSON.stringify({
		type: "stt_cancel",
		clip_id: clip.id,
		generation: clip.epoch,
	});
}

export function clipHeader(clip: {
	id: string;
	mime: string;
	epoch: number;
}): string {
	return JSON.stringify({
		type: "clip",
		id: clip.id,
		mime: clip.mime,
		generation: clip.epoch,
	});
}

/// A turn the caller typed instead of spoke.
///
/// It is a transcript that needs no transcription, and it carries the same
/// epoch guard as a clip: the epoch the browser held when the caller sent it.
/// The server drops it if a transfer has moved the epoch on since, and echoes
/// it back as a `transcript` frame with the same id when it is taken.
export function typedTurnMessage(turn: {
	id: string;
	epoch: number;
	text: string;
}): string {
	return JSON.stringify({
		type: "typed_turn",
		id: turn.id,
		generation: turn.epoch,
		text: turn.text,
	});
}

export async function postJson(
	url: string,
	body: Record<string, unknown>,
): Promise<Record<string, unknown>> {
	const response = await fetch(url, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	});
	if (!response.ok) throw new Error(`HTTP ${response.status}`);
	return (await response.json()) as Record<string, unknown>;
}

File: apps/backend/src/protocol.rs
//! The WebSocket protocol from the service to the browser: every message the
//! page can receive, one variant per `type`.
//!
//! The browser's half is `ServerMessage` in `apps/frontend/src/protocol.ts`,
//! beside the commands the page sends this way. Both halves are held to the
//! examples in `apps/frontend/tests/fixtures/server-messages.json`: each one
//! must serialize back to itself here and decode to itself there, and a
//! variant without an example fails on both sides.
use crate::history::TranscriptEntry;
use serde::Serialize;
use serde_json::{Map, Value};

#[derive(Clone, Debug, PartialEq, Serialize)]
#[cfg_attr(test, derive(serde::Deserialize))]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    /// The answer to `hello`: the protocol version, and which optional
    /// transports this socket uses.
    HelloAck {
        version: u64,
        stt_streaming: bool,
        audio_streaming: bool,
        mse_mp3: bool,
    },
    /// The answer to `ping`, echoing its `nonce` and `time` as they came, or
    /// null for one it did not carry.
    Pong { nonce: Value, time: Value },
    /// The turn epoch. Sent in every connection's snapshot and whenever a
    /// rescue or a new leg moves it; the browser stamps clips with it.
    Epoch { generation: u64 },
    /// A transfer began starting a leg on `route`. Clips recorded until the
    /// next `epoch` are addressed to that leg.
    Candidate { route: String, generation: u64 },
    /// The candidate leg was adopted, rolled back, or rescued away.
    CandidateCleared { generation: u64 },
    /// The line's status: who is on it, their model and thinking level, and
    /// which of those the caller may change. Carried as the PBX and the
    /// coordinator build it (see [`ServerMessage::status`]); its fields are
    /// typed in the browser and pinned by the fixture, and are typed here
    /// once the status projection itself is (#60).
    Status(Map<String, Value>),
    /// A clip was taken: whole, or opened as a stream (`streaming`).
    Accepted {
        id: String,
        #[serde(default, skip_serializing_if = "std::ops::Not::not")]
        streaming: bool,
    },
    /// A streaming clip cannot continue as a stream; the browser sends it
    /// whole instead.
    Abandoned { id: String, reason: String },
    /// The final transcript of a clip, or the echo of a typed turn.
    Transcript { id: String, text: String },
    /// A transcript was steered into the turn in progress, or queued behind
    /// `waiting` turns.
    Queued {
        id: String,
        waiting: u64,
        steered: bool,
    },
    /// Something the caller should see went wrong. `id` names the clip or
    /// typed turn it answers, when there is one.
    Error {
        #[serde(default, skip_serializing_if = "Option::is_none")]
        id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        code: Option<ErrorCode>,
        message: String,
    },
    /// A turn was dispatched to the leg on `route`.
    Thinking { route: String, waiting: u64 },
    /// A tool call on the live leg started or ended (`state` is `start` or
    /// `end`); `label` names the leg.
    Activity {
        state: String,
        tool: String,
        detail: String,
        label: String,
    },
    /// A turn settled with this written reply.
    Reply { text: String, route: String },
    /// A line was spoken to the caller and kept in the transcript.
    Spoken { entry: TranscriptEntry },
    /// The transcript so far, in every connection's snapshot.
    History { entries: Vec<TranscriptEntry> },
    /// An utterance of synthesized speech begins; its audio follows as
    /// binary frames.
    AudioStart {
        generation: u64,
        sequence: u64,
        mime: String,
        format: String,
    },
    /// The utterance `sequence` has no more audio.
    AudioDone {
        generation: u64,
        sequence: u64,
        done: bool,
    },
    /// All audio for the reply to `response_id` has been sent; `success` is
    /// false when some of it could not be synthesized.
    FinalResponseAudioClosed {
        response_id: String,
        generation: u64,
        success: bool,
    },
    /// A display action for the stage. `action` is validated and normalized
    /// by `visual_protocol` and pinned by `display-actions.json`. `seq` is the
    /// delivery sequence the browser confirms it by: the socket writer stamps
    /// it on a live action (`stamp_display_seq`), and a snapshot replay
    /// carries the watermark.
    Display {
        action: Value,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        seq: Option<u64>,
    },
    /// The agent asked for a view of the workspace.
    View { target: String, reason: String },
    /// The browser's `screen_state` report was applied.
    ScreenStateAck,
}

/// Why an `error` needs handling beyond showing its message.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[cfg_attr(test, derive(serde::Deserialize))]
#[serde(rename_all = "snake_case")]
pub enum ErrorCode {
    /// The clip or typed turn was recorded for a leg the call has since left;
    /// the browser drops it rather than retrying.
    StaleEpoch,
}

impl ServerMessage {
    /// The status message for a status projection, from
    /// `Coordinator::status_json` or `Switchboard::status`. The projection
    /// names its own `type`, which the tag replaces.
    pub fn status(projection: Value) -> Self {
        let mut fields = match projection {
            Value::Object(fields) => fields,
            _ => Map::new(),
        };
        fields.remove("type");
        Self::Status(fields)
    }

    /// An `error` that answers no particular clip or turn.
    pub fn error(message: impl Into<String>) -> Self {
        Self::Error {
            id: None,
            code: None,
            message: message.into(),
        }
    }

    /// An `error` that answers the clip or typed turn `id`.
    pub fn error_for(id: impl Into<String>, message: impl Into<String>) -> Self {
        Self::Error {
            id: Some(id.into()),
            code: None,
            message: message.into(),
        }
    }

    /// The message as the JSON the browser receives.
    pub fn to_value(&self) -> Value {
        serde_json::to_value(self)
            .expect("a server message holds only string-keyed JSON, so it always serializes")
    }
}

#[cfg(test)]
#[path = "../tests/test_protocol.rs"]
mod tests;
